ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -429,24 +429,69 @@ func (r *RayJobReconciler) getOrCreateRayClusterInstance(ctx context.Context, ra
+
+		// Case2: In-tree autoscaling is enabled, only the autoscaler should update replicas to prevent race conditions
+		// between user updates and autoscaler decisions. RayJob controller should not modify the replica. Consider this scenario:
+		// 1. The autoscaler updates replicas to 10 based on the current workload.
+		// 2. The user updates replicas to 15 in the RayJob YAML file.
+		// 3. Both RayJob controller and the autoscaler attempt to update replicas, causing worker pods to be repeatedly created and terminated.
+		if rayJobInstance.Spec.RayClusterSpec.EnableInTreeAutoscaling != nil && *rayJobInstance.Spec.RayClusterSpec.EnableInTreeAutoscaling {
+			// Note, currently, there is no method to verify if the user has updated the RayJob since the last reconcile.
+			// In future, we could utilize annotation that stores the hash of the RayJob since last reconcile to compare.
+			// For now, we just log a warning message to remind the user regadless whether user has updated RayJob.
+			r.Log.Info("Since in-tree autoscaling is enabled, any adjustments made to the RayJob will be disregarded and will not be propagated to the RayCluster.")
+			return rayClusterInstance, nil
+		}
+
+		// Case3: In-tree autoscaling is disabled, respect the user's replicas setting.
+		// Loop over all worker groups and update replicas.
+		areReplicasIdentical := true
+		for i := range rayJobInstance.Spec.RayClusterSpec.WorkerGroupSpecs {
+			if *rayClusterInstance.Spec.WorkerGroupSpecs[i].Replicas != *rayJobInstance.Spec.RayClusterSpec.WorkerGroupSpecs[i].Replicas {
+				areReplicasIdentical = false
+				*rayClusterInstance.Spec.WorkerGroupSpecs[i].Replicas = *rayJobInstance.Spec.RayClusterSpec.WorkerGroupSpecs[i].Replicas
+			}
+		}
+
+		// Other specs rather than replicas are changed, warn the user that the RayJob supports replica changes only.
+		if !utils.CompareJsonStruct(rayClusterInstance.Spec, *rayJobInstance.Spec.RayClusterSpec) {
+			r.Log.Info("RayJob supports replica changes only. Adjustments made to other specs will be disregarded as they may cause unexpected behavior")
+		}
+
+		// Avoid updating the RayCluster's replicas if it's identical to the RayJob's replicas.
+		if areReplicasIdentical {
 			return rayClusterInstance, nil
 		}
-		rayClusterInstance.Spec = *rayJobInstance.Spec.RayClusterSpec
 
-		r.Log.Info("Update ray cluster spec", "raycluster", rayClusterNamespacedName)
+		r.Log.Info("Update ray cluster replica", "raycluster", rayClusterNamespacedName)
 		if err := r.Update(ctx, rayClusterInstance); err != nil {
-			r.Log.Error(err, "Fail to update ray cluster instance!", "rayCluster", rayClusterNamespacedName)
+			r.Log.Error(err, "Fail to update ray cluster replica!", "rayCluster", rayClusterNamespacedName)
 			// Error updating the RayCluster object.
 			return nil, client.IgnoreNotFound(err)
 		}
+
 	} else if errors.IsNotFound(err) {
+		if len(rayJobInstance.Spec.ClusterSelector) == 0 && rayJobInstance.Spec.RayClusterSpec == nil {
```

Comment:
This is not good. We should not try to `GET` RayCluster if both `ClusterSelector` and `RayClusterSpec` are not set. Please add a comment "TODO" and a new issue to track the progress.